### ISSUE TITLE
复现不出结果

### ISSUE THREAD

**Issue by yumulinfeng1** [OFA-Sys/Chinese-CLIP#28]
你好
我跑MUGE数据集，看训练过程中的验证集指标一直没有变化如下：
2022-12-08,21:22:28 | INFO | Rank 1 | Validation Result (epoch 2 @ 4650 steps) | Valid Loss: 1.668444 | Image2Text Acc: 32.41 | Text2Image Acc: 32.94 | logit_scale: 4.595 | Valid Batch Size: 48
2022-12-08,21:22:28 | INFO | Rank 0 | Validation Result (epoch 2 @ 4650 steps) | Valid Loss: 1.668444 | Image2Text Acc: 32.41 | Text2Image Acc: 32.94 | logit_scale: 4.595 | Valid Batch Size: 48

准确率一直是32左右。
请问怎么浮现出仓库中写的60+的准确率？

**Comment by yumulinfeng1** [comment]
> @yumulinfeng1 您好，有以下几个要点请您注意下：
> 
> 1. 训练过程中进行的验证，都是在一个batch内部计算Acc，这里计算的结果和最终全局召回的Recall@1/5/10及Mean Recall不是一个指标，仅用于训练过程判断收敛趋势。如果要得到可以和我们汇报的Recall@1/5/10以及Mean Recall对比的指标，请按照我们readme中[跨模态检索](https://github.com/OFA-Sys/Chinese-CLIP#%E8%B7%A8%E6%A8%A1%E6%80%81%E6%A3%80%E7%B4%A2)部分，描述的训练→特征提取→KNN召回→计算Recall这个过程，完整走一遍finetune和测试集全图片池召回的过程。
> 2. finetune的效果与超参数也有关系。能够跑出最优结果的超参数，请参见我们技术报告的附录部分A.3，给出了每个规模、每个数据集的最优超参数，供您参考。
> 3. 请保证预训练ckpt有正常load进来。
> 
> 还不太清楚您所说的60+具体是哪个指标。**建议您可以先尝试对齐zero-shot结果，无须finetune。直接用想要对齐规模的模型预训练ckpt、和我们提供的预处理好的数据集，走完特征提取→KNN召回→计算Recall这个流程，看下Recall指标能否和我们汇报的结果一致，也供您熟悉一下用Chinese-CLIP进行图文检索的一个标准流程。** 在此基础上，进一步跑个finetune，观察下效果能否提升。

你好 请问一下你仓库里给的验证集的log示例：
2022-06-16,11:06:00 | INFO | Rank 0 | Validation Result (epoch 1 @ 150 steps) | Valid Loss: 0.503617 | Image2Text Acc: 84.76 | Text2Image Acc: 84.37 | logit_scale: 4.605 | Valid Batch Size: 128


是哪个数据集的验证集？
如果是MUGE，那验证集的差距太大了。
我这边跑MUGE验证集的准确率就没变过，一直在32左右，是否有问题

**Comment by yangapku** [comment]
您好，这个应该是来自MUGE的log，您可以贴一下您实际的运行脚本和log，便于我们判断

**Comment by yumulinfeng1** [comment]
脚本如下：
# Number of GPUs per GPU worker
GPUS_PER_NODE=2 
# Number of GPU workers, for single-worker training, please set to 1
WORKER_CNT=1
# The ip address of the rank-0 worker, for single-worker training, please set to localhost
export MASTER_ADDR=10.5.75.116
# The port for communication
export MASTER_PORT=8514
# The rank of this worker, should be in {0, ..., WORKER_CNT-1}, for single-worker training, please set to 0
export RANK=0 

export DATAPATH=/xxx/Text_Based_Image_Retrieval/data
export OUTLOGPATH=/xxx/Text_Based_Image_Retrieval/Chinese-CLIP-master/experiments


# data options
train_data=${DATAPATH}/datasets/MUGE/lmdb/train
val_data=${DATAPATH}/datasets/MUGE/lmdb/valid # if val_data is not specified, the validation will be automatically disabled

# restore options
resume=${OUTLOGPATH}/pretrained_weights/clip_cn_vit-b-16.pt # or specify your customed ckpt path to resume
reset_data_offset="--reset-data-offset"
reset_optimizer="--reset-optimizer"
# reset_optimizer=""

# output options
output_base_dir=${OUTLOGPATH}/
name=muge_finetune_vit-b-16_roberta-base_bs32
save_step_frequency=999999 # disable it
save_epoch_frequency=1
log_interval=1
report_training_batch_acc="--report-training-batch-acc"
# report_training_batch_acc=""

# training hyper-params
context_length=52
warmup=100
batch_size=48
valid_batch_size=48
lr=2e-5
wd=0.001
max_epochs=16
valid_step_interval=150
valid_epoch_interval=1
vision_model=ViT-B-16
text_model=RoBERTa-wwm-ext-base-chinese
use_augment="--use-augment"
# use_augment=""

python3 -m torch.distributed.launch --nproc_per_node=${GPUS_PER_NODE} --nnodes=${WORKER_CNT} --node_rank=${RANK} \
          --master_addr=${MASTER_ADDR} --master_port=${MASTER_PORT} cn_clip/training/main.py \
          --train-data=${train_data} \
          --val-data=${val_data} \
          --resume=${resume} \
          ${reset_data_offset} \
          ${reset_optimizer} \
          --logs=${output_base_dir} \
          --name=${name} \
          --save-step-frequency=${save_step_frequency} \
          --save-epoch-frequency=${save_epoch_frequency} \
          --log-interval=${log_interval} \
          ${report_training_batch_acc} \
          --context-length=${context_length} \
          --warmup=${warmup} \
          --batch-size=${batch_size} \
          --valid-batch-size=${valid_batch_size} \
          --valid-step-interval=${valid_step_interval} \
          --valid-epoch-interval=${valid_epoch_interval} \
          --lr=${lr} \
          --wd=${wd} \
          --max-epochs=${max_epochs} \
          --vision-model=${vision_model} \
          ${use_augment} \
          --text-model=${text_model}




log如下：

2022-12-09,11:31:47 | INFO | Rank 0 | Global Steps: 25500/41728 | Train Epoch: 10 [194688/250368 (78%)] | Loss: 0.191398 | Image2Text Acc: 92.71 | Text2Image Acc: 94.79 | Data Time: 0.060s | Batch Time: 0.835s | LR: 0.000007 | logit_scale: 4.584 | Global Batch Size: 96
2022-12-09,11:31:47 | INFO | Rank 0 | Begin to eval on validation set (epoch 10 @ 25500 steps)...
2022-12-09,11:32:56 | INFO | Rank 1 | Evaluated 100/319 batches...
2022-12-09,11:32:58 | INFO | Rank 0 | Evaluated 100/319 batches...
2022-12-09,11:34:04 | INFO | Rank 1 | Evaluated 200/319 batches...
2022-12-09,11:34:10 | INFO | Rank 0 | Evaluated 200/319 batches...2022-12-09,11:35:13 | INFO | Rank 1 | Evaluated 300/319 batches... 0
2022-12-09,11:35:21 | INFO | Rank 0 | Evaluated 300/319 batches...
2022-12-09,11:35:35 | INFO | Rank 1 | Validation Result (epoch 10 @ 25500 steps) | Valid Loss: 2.025491 | Image2Text Acc: 32.39 | Text2Image Acc: 33.03 | logit_scale: 4.584 | Valid Batch Size: 48
2022-12-09,11:35:35 | INFO | Rank 0 | Validation Result (epoch 10 @ 25500 steps) | Valid Loss: 2.025491 | Image2Text Acc: 32.39 | Text2Image Acc: 33.03 | logit_scale: 4.584 | Valid Batch Size: 48

**Comment by yangapku** [comment]
请问方便提供一个完整的训练log文件链接吗？我们初步判断，是没有正确load进来预训练的ckpt参数，建议您也检查下ckpt是否放在`--resume`指定的位置。

**Comment by yangapku** [comment]
您也可以看下log中是否有"=> no checkpoint found at"这样的日志打印出来，确认这一点。

**Comment by yangapku** [comment]
收到，我们也去排查下

**Comment by yumulinfeng1** [comment]
> 2022-12-08,18:15:41 | INFO | Rank 0 | => loaded checkpoint '/storage1/xxx/Text_Based_Image_Retrieval/Chinese-CLIP-master/experiments/pretrained_weights/clip_cn_vit-b-16.pt' (epoch 15 @ 0

好的谢谢。麻烦排查下。从训练日志看，训练记得loss一直在下降。训练接acc一直在提升。  但是验证集的的loss和acc从一开始就没怎么变过，acc维持在32左右浮动

**Comment by yangapku** [comment]
@yumulinfeng1 您好，关于您在issue提出的问题，大概有以下几点结论和建议：

1. 关于您所提到的finetune时，验证集inbatch acc一直较低的问题，主要是因为MUGE数据集验证集存在文对图一对多。在默认验证集不shuffle、且GPU分布式卡数较少的情况下，一个验证batch里面会有多个图文样本的文本相同的情况。由于我们的val inbatch acc是最简单的实现版本，只把样本自己作为ground truth，所以这种情况下会出现inbatch acc计算结果偏低的现象。**实际上模型训练和收敛完全没有影响，最终走完整个评测流程，得到的Recall指标如果你有尝试跑过，应该也没问题。** 更具体的描述参见PR #29 ，我们也做了个小改动，把finetune时验证集的sampler也改成shuffle的情况了，规避掉这种特殊情况。**现在如果拉取最新代码再跑，不会出现验证集inbatch acc一直较低这个情况了。**
2. 关于模型finetune的效果对比，建议您和我们汇报结果所对比的，应该是跑完图文检索的评测流程最终得到的Recall（或者同一个零样本图片分类数据集上的准确率），而不是对比验证集inbatch acc。这是因为验证集inbatch acc显然和`valid_batch_size`有关，您所跑的`valid_batch_size=48`和我们脚本默认的`valid_batch_size=128`也不相同，所以和我们那个log样例写的inbatch acc对比意义不大，或者说您最新代码跑出来一定会比我们的inbatch acc要高，因为batch小，负例更少。验证集inbatch acc只用于您自己几组实验评估收敛趋势。
3. 您finetune使用的训练batch size很小，2卡总计96。根据我们Readme中的描述，对比学习的训练收敛和稳定性和总batch size相关。如您使用更小的batch size（相比脚本默认配置128 per-GPU * 8 GPU），建议使用更小的学习率。您的学习率为2e-5，我们实测这组超参下，finetune最终的Recall甚至比不finetune的zeroshot结果还会要低（MUGE Mean Recall (MR) 68左右，对比我们base规模预训练模型zeroshot的MR 71.1）。建议您试试更小的学习率或者更大的batch（可以考虑用Readme中的重计算策略）。我们有尝试如果您提供这组超参，学习率降到2e-6，最终的MR将达到75.5以上，相比于zeroshot的MR 71.1会有提升。供您参考。

如果有更多问题，欢迎继续留言。如果觉得Chinese-CLIP代码库对您有帮助，请您为我们点点star⭐️并推荐给身边的朋友们！

**Comment by 6Roy** [comment]
请问您使用的torch版本是多少？我在复现的时候一直出现问题
